VoiceOver/DataModules.lua

```diff
@@ -0,0 +1,154 @@
+            if IsAddOnLoadOnDemand(name) then
+                LoadAddOn(name)
+            end
+        end
+    end
+end
+
+function DataModules:GetNPCGossipTextHash(soundData)
+    local npcId = VoiceOverUtils:getIdFromGuid(soundData.unitGuid)
+    local text = soundData.text
+
+    local text_entries = {}
+
+    for _, module in self:GetModules() do
+        local data = module.NPCToTextToTemplateHash
+        if data then
+            local npc_gossip_table = data[npcId]
+            if npc_gossip_table then
+                for text, hash in pairs(npc_gossip_table) do
+                    text_entries[text] = text_entries[text] or hash -- Respect module priority, don't overwrite the entry if there is already one
+                end
+            end
+        end
+    end
+
+    local best_result = fuzzySearchBestKeys(text, text_entries)
+    return best_result and best_result.value
+end
+
+function DataModules:GetQuestLogNPCID(questId)
+    for _, module in self:GetModules() do
+        local data = module.QuestlogNpcGuidTable
+        if data then
+            local npcId = data[questId]
+            if npcId then
+                return npcId
+            end
+        end
+    end
+end
+
+function DataModules:GetQuestIDByQuestTextHash(hash, npcId)
+    local hashWithNpc = format("%s:%d", hash, npcId)
+    for _, module in self:GetModules() do
+        local data = module.QuestTextHashToQuestID
+        if data then
+            local questId = data[hashWithNpc] or data[hash]
+            if questId then
+                return questId
+            end
+        end
+    end
+end
+
+local getFileNameForEvent =
+{
+    accept = function(soundData) return format("%d-%s", soundData.questId, "accept") end,
+    progress = function(soundData) return format("%d-%s", soundData.questId, "progress") end,
+    complete = function(soundData) return format("%d-%s", soundData.questId, "complete") end,
+    gossip = function(soundData) return DataModules:GetNPCGossipTextHash(soundData) end,
+}
+
+function DataModules:PrepareSound(soundData)
+    soundData.fileName = getFileNameForEvent[soundData.event](soundData) or error([[Unhandled VoiceOver sound event "%s"]], soundData.event)
+    for _, module in self:GetModules() do
+        local data = module.SoundLengthTable
+        if data then
+            local length = data[soundData.fileName]
+            if length then
+                soundData.filePath = format([[Interface\AddOns\%s\%s]], module.MODULE_NAME, module.GetSoundPath and module:GetSoundPath(soundData.fileName, soundData.event) or soundData.fileName)
+                soundData.length = length
+                soundData.module = module
+                return true
+            end
+        end
+    end
+    soundData.fileName = "missingSound" -- Not sure why this is needed, the presence of the sound file can be checked by whether it has a sound length record
```

Comment:
this was originally done to have a key to debounce the addition of a sound but no longer makes sense given how the addon has developed. We could probably remove.

Reply:
I'm going to leave it as is for now, it's still used in that debounce timer, and if we move the "will play" check out of the timer - if I'm understanding the purpose of "debounce" correctly, the message "Sound does not exist for:" might show up in chat multiple times. And since the debounce timer uses fileName as key, we can't have a nil key.